### PR TITLE
Readers&Writers

### PR DESCRIPTION
**Pull request by radex**

### CODE REVIEW

**Comment by radex** [comment]
@michalpopek Don't get scared, there's a lot of noise in this diff in tests and docs - you can skim through that, the important part is Database/WorkQueue and Database/test

File: docs-master/Model.md [hunk 1]
```diff
@@ -93,25 +95,43 @@ class Post extends Model {
 }
 ```
 
-## Relation fields
+### Compound fields
```
Comment (michalpopek) [inline docs-master/Model.md:98]: I think this `computed` is a widely recognizable term for this (Vue, MobX), although we must note the they aren't reactive in any way. So maybe:

```suggestion
### Derived fields
```

**Review by michalpopek — COMMENTED** [review]
@radex went through this PR and LGTM. Left some remarks about the docs part. I'll merge it, once we make a decision about the reader API in https://github.com/Nozbe/WatermelonDB/issues/1015

File: src/Database/index.d.ts [hunk 2]
```diff
@@ -11,8 +11,12 @@ declare module '@nozbe/watermelondb/Database' {
...
   import { Observable } from 'rxjs'
 
-  export interface ActionInterface {
-    subAction<T>(action: () => Promise<T>): Promise<T>
+  interface ReaderInterface {
+    callReader<T>(work: () => Promise<T>): Promise<T>
+  }
+  interface WriterInterface extends ReaderInterface {
+    callWriter<T>(work: () => Promise<T>): Promise<T>
+    batch(...records: (Model | null | void | false | Promise<void>)[]): Promise<void>
```
[the remark this answers is not in the document]
Comment (radex) [reply src/Database/index.d.ts:19]: hmm, this is copy&pasted from Database/index.d.ts

[the remark this answers is not in the document]
Comment (michalpopek) [reply src/Database/index.d.ts:19]: My bad, wasn't aware that `void` and `undefined` are interchangeable in this case

**Comment by radex** [comment]
@michalpopek I say go for merge if you don't have any more feedback this. I still consider #1015 to be open, but at this point, it'll be easier for me to evolve the API once merged. We're not shipping next minor version soon anyway

**Pull request merged**